src/main/kotlin/com/frcteam3636/frc2025/subsystems/drivetrain/AbsolutePose.kt

```diff
@@ -51,47 +51,111 @@ class AbsolutePoseProviderInputs : LoggableInputs {
+    }
 }
 
-class LimelightPoseProvider(private val name: String) : AbsolutePoseProvider {
-    override fun updateInputs(inputs: AbsolutePoseProviderInputs) {
-        // This is mostly pulled from the LimeLight docs.
-        // https://docs.limelightvision.io/docs/docs-limelight/apis/limelight-lib#4-field-localization-with-megatag
-        val estimate = LimelightHelpers.getBotPoseEstimate_wpiBlue(name)
-        val trackedTag = LimelightHelpers.getTargetPose3d_RobotSpace(name)
-
-        // Only trust the measurement if we see multiple tags.
-        inputs.measurement = if (estimate != null && estimate.tagCount >= 2) {
-            AbsolutePoseMeasurement(
-                estimate.pose,
-                Seconds.of(estimate.timestampSeconds),
-                VecBuilder.fill(0.7, 0.7, 9999999.0),
-            )
-        } else {
-            null
+class LimelightPoseProvider(
+    private val name: String,
+    private val algorithm: LimelightAlgorithm,
+) : AbsolutePoseProvider {
+    // References:
+    // https://docs.limelightvision.io/docs/docs-limelight/tutorials/tutorial-swerve-pose-estimation
+    // https://docs.limelightvision.io/docs/docs-limelight/apis/limelight-lib#4-field-localization-with-megatag
+
+    private fun updateCurrentMeasurement(): AbsolutePoseMeasurement? {
+        return when (algorithm) {
+            is LimelightAlgorithm.MegaTag ->
+                LimelightHelpers.getBotPoseEstimate_wpiBlue(name)?.let { estimate ->
+                    // Reject zero tag or low-quality one tag readings
+                    if (estimate.tagCount == 0) return null
+                    if (estimate.tagCount == 1) {
+                        val fiducial = estimate.rawFiducials[0]
+                        if (fiducial == null
+                            || fiducial.ambiguity > AMBIGUITY_THRESHOLD
+                            || fiducial.distToCamera > MAX_SINGLE_TAG_DISTANCE
+                        ) return null
+                    }
+
+                    return AbsolutePoseMeasurement(
+                        estimate.pose,
+                        Seconds.of(estimate.timestampSeconds),
+                        // This value is pulled directly from the Limelight docs (linked at the top of this class)
+                        VecBuilder.fill(.5, .5, 9999999.0)
+                    )
+                }
+
+            is LimelightAlgorithm.MegaTag2 -> {
+                LimelightHelpers.SetRobotOrientation(
+                    name,
+                    algorithm.gyroPosition.degrees,
+                    // The Limelight sample code leaves these as zero, and the API docs call them "Unnecessary."
+                    0.0, 0.0, 0.0, 0.0, 0.0
+                )
+
+                LimelightHelpers.getBotPoseEstimate_wpiBlue_MegaTag2(name)?.let { estimate ->
+                    if (algorithm.gyroVelocity.abs(DegreesPerSecond) > 720.0) return null
+                    if (estimate.tagCount == 0) return null
```

Comment:
Same thing, Limelight helpers handle this for you. https://github.com/LimelightVision/limelightlib-wpijava/blob/main/LimelightHelpers.java#L679